### ISSUE TITLE
Composer\Script\Event should have access to originating event details

### ISSUE THREAD

**Issue by freality** [composer/composer#8183]
When `Composer\EventDispatcher\EventDispatcher::doDispatch()` encounters a custom script, it creates a new `Composer\Script\Event` which loses any `InstallerEvent, PackageEvent, etc.` details.

Adding `getOriginatingEvent` and `setOriginatingEvent` to `Composer\Script\Event` would allow handlers access to the originating details without modifying the current constructor.

Example `composer.json`:
```json
"scripts": {
        "pre-install-cmd": [
            "DrupalProject\\composer\\ScriptHandler::checkComposerVersion"
        ],
        "pre-update-cmd": [
            "DrupalProject\\composer\\ScriptHandler::checkComposerVersion"
        ],
        "post-install-cmd": [
            "DrupalProject\\composer\\ScriptHandler::createRequiredFiles"
        ],
        "post-update-cmd": [
            "DrupalProject\\composer\\ScriptHandler::createRequiredFiles"
        ],
        "post-package-install": [
            "Drush\\composer\\DrushHandler::getDetails",
            "@details | xargs echo"
        ],
        "pre-package-uninstall": [
            "@details | grep 'packageName' | xargs ./bin/disable-cms-package"
        ],
        "details": "My\\composer\\Handler::getPackageDetails"
    }
```

**Comment by alcohol** [comment]
You mean, that because you call `details` from within `pre-package-uninstall`, that you expect Composer to let `details` know that it is being called from `pre-package-uninstall` ?

**Comment by freality** [comment]
Yes. 

I'd like the ability to chain composer information from the operation to other scripts. From what I see, this is not currently possible.

If my proposed solution is acceptable, I can create the PR.

**Comment by alcohol** [comment]
Eh, I would say no. But check with @Seldaek

**Comment by Seldaek** [comment]
The request sounds somewhat reasonable although the way you are using it with mixing `@details` and shell notation is not going to work.. 

Also not sure what happens with setOriginatingEvent in the case of a longer chain of `@foo` pointing to `@bar` etc..  I guess you can just call getOriginatingEvent until you have nothing left, or probably in doDispatch it should call getOriginatingEvent to make sure it resolves the originating event set on the Script\Event instance to the real one.

**Comment by freality** [comment]
> Also not sure what happens with setOriginatingEvent in the case of a longer chain of `@foo` pointing to `@bar` etc.. I guess you can just call getOriginatingEvent until you have nothing left, or probably in doDispatch it should call getOriginatingEvent to make sure it resolves the originating event set on the Script\Event instance to the real one.

If you're talking about chaining within a single listener/callable (e.g. `@foo @bar @baz`), I don't see anything in the code that supports that. If someone were to attempt this the method associated with `@foo` would receive an event where `$event->getArguments() == ['@bar', '@baz']`.

**Comment by Seldaek** [comment]
Nope I mean if you chain to another script which chains to another which
chains to another

If that ain't clear I'll spell it out tomorrow on my laptop

>

**Comment by alcohol** [comment]
I think @Seldaek meant

```
"scripts": {
  "foo": "@bar",
  "bar": "@baz",
  "baz": "echo"
}
```

**Comment by Seldaek** [comment]
Yup that's what I meant, didn't feel like typing it on a phone though ;)

**Comment by freality** [comment]
First, I'd like to thank you all for considering and evaluating this request.

I didn't think of that style of chaining. But generally, I believe that API's should make best effort to respond as expected. In this case, `getOriginatingEvent` should return the original event.

I'd implement this in `setOriginatingEvent` to keep the logic encapsulated in the implementor.

Like:
```php
    /**
     * Set the originating event.
     *
     * @param \Composer\EventDispatcher\Event $event
     * @return $this
     */
    public function setOriginatingEvent(BaseEvent $event)
    {
      $this->originatingEvent = $this->calculateOriginatingEvent($event);

      return $this;
    }

    /**
     * Returns the upper-most event in chain.
     *
     * @param \Composer\EventDispatcher\Event $event
     * @return \Composer\EventDispatcher\Event
     */
    private function calculateOriginatingEvent(BaseEvent $event)
    {
      if ($event instanceof Event && $event->getOriginatingEvent()) {
        return $this->calculateOriginatingEvent($event->getOriginatingEvent());
      }

      return $event;
    }
```

**Comment by Seldaek** [comment]
Sure that sounds fine to me.

**Comment by freality** [comment]
The referenced PR #8186 implements the discussed changes. I'm not sure why the AppVeyor test build is failing. The failed test `Composer\Test\Util\FilesystemTest::testJunctions` isn't related. If there's anything I should do to right the test, please advise.

**Comment by alcohol** [comment]
The junction tests fail quite regularly. You can ignore those failures.

**Comment by Seldaek** [comment]
Closed by #8186
